static/app/components/events/highlights/editHighlightsModal.tsx

```diff
@@ -0,0 +1,520 @@
+  const contextColumns: React.ReactNode[] = [];
+  for (let i = 0; i < ctxItems.length; i += ctxColumnSize) {
+    contextColumns.push(
+      <EditHighlightColumn key={`ctx-column-${i}`}>
+        {ctxItems.slice(i, i + ctxColumnSize).map(([contextType, contextKeys], j) => (
+          <EditContextContainer key={`ctxv-item-${i}-${j}`}>
+            <ContextType>{contextType}</ContextType>
+            {contextKeys.map((contextKey, k) => {
+              const isDisabled = ctxDisableMap[contextType]?.has(contextKey) ?? false;
+              return (
+                <Fragment key={`ctx-key-${i}-${j}-${k}`}>
+                  <EditButton
+                    aria-label={`Add ${contextKey} from ${contextType} context to highlights`}
+                    icon={<IconAdd />}
+                    size="xs"
+                    onClick={() => onAddContextKey(contextType, contextKey)}
+                    disabled={isDisabled}
+                    title={isDisabled && t('Already highlighted')}
+                    tooltipProps={{delay: 500}}
+                  />
+                  <HighlightKey disabled={isDisabled} aria-disabled={isDisabled}>
+                    {contextKey}
+                  </HighlightKey>
+                </Fragment>
+              );
+            })}
+          </EditContextContainer>
+        ))}
+      </EditHighlightColumn>
+    );
+  }
+
+  return (
+    <EditHighlightSection {...props}>
+      <Subtitle>{t('Context')}</Subtitle>
+      <EditHighlightSectionContent columnCount={columnCount}>
+        {contextColumns}
+      </EditHighlightSectionContent>
+    </EditHighlightSection>
+  );
+}
+
+export default function EditHighlightsModal({
+  Header,
+  Body,
+  Footer,
+  event,
+  highlightContext: prevHighlightContext,
+  highlightTags: prevHighlightTags,
+  highlightPreset,
+  project,
+  closeModal,
+}: EditHighlightsModalProps) {
+  const [highlightContext, setHighlightContext] =
+    useState<HighlightContext>(prevHighlightContext);
+  const [highlightTags, setHighlightTags] = useState<HighlightTags>(prevHighlightTags);
+
+  const organization = useOrganization();
+  const api = useApi();
+  const queryClient = useQueryClient();
+
+  const {mutate: saveHighlights, isLoading} = useMutation<Project, RequestError>({
+    mutationFn: () => {
+      return api.requestPromise(`/projects/${organization.slug}/${project.slug}/`, {
+        method: 'PUT',
+        data: {
+          highlightContext,
+          highlightTags,
+        },
+      });
+    },
+    onSuccess: (_updatedProject: Project) => {
+      addSuccessMessage(
+        tct(`Successfully updated highlights for '[projectName]' project`, {
+          projectName: project.name,
+        })
+      );
+      setApiQueryData<Project>(
+        queryClient,
+        makeDetailedProjectQueryKey({
+          orgSlug: organization.slug,
+          projectSlug: project.slug,
+        }),
+        data => (data ? {...data, highlightTags, highlightContext} : data)
```

Comment:
nit: usually better to use the data in the callback (`updatedProject`) since we know that's the most up to date server state